app/controllers/accounts_controller.rb

```diff
@@ -23,7 +23,22 @@ def create
 
   private
 
+  memoize def public_shared_accounts
+    Set.new(AccountShare.accepted.for_public.pluck(:account_id))
+  end
+
+  memoize def private_shared_accounts
+    Set.new(AccountShare.accepted.for(current_user).pluck(:account_id))
+  end
+
   helper_method memoize def account
     Account.visible_for(current_user).find(ps.fetch(:id))
   end
+
+  helper_method def public_share?(account)
+    return false if current_user.account.id == account.parent.id ||
```

Comment:
```suggestion
    return false if current_user.account_id == account.parent_id ||
```

this way it will be faster